pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3IndexFilterOperator.java

```diff
@@ -90,110 +92,139 @@ public H3IndexFilterOperator(IndexSegment segment, Predicate predicate, int numD
-    partialMatchH3Ids.removeAll(fullMatchH3Ids);
-    MutableRoaringBitmap partialMatchDocIds = new MutableRoaringBitmap();
-    for (long partialMatchH3Id : partialMatchH3Ids) {
-      partialMatchDocIds.or(_h3IndexReader.getDocIds(partialMatchH3Id));
-    }
-
-    return getFilterBlock(fullMatchDocIds, partialMatchDocIds);
   }
 
   /**
    * Returns the H3 ids that is ALWAYS fully covered by the circle with the given distance as the radius and a point
    * within the _h3Id hexagon as the center.
+   * <p>The farthest distance from the center of the center hexagon to the center of a hexagon in the nth ring is
+   * {@code sqrt(3) * n * edgeLength}. Counting the distance from the center to a point in the hexagon, which is up
+   * to the edge length, it is guaranteed that the hexagons in the nth ring are always fully covered if:
+   * {@code distance >= (sqrt(3) * n + 2) * edgeLength}.
    */
-  private List<Long> getFullMatchH3Ids(double distance) {
+  private List<Long> getAlwaysMatchH3Ids(double distance) {
     // NOTE: Pick a constant slightly larger than sqrt(3) to be conservative
     int numRings = (int) Math.floor((distance / _edgeLength - 2) / 1.7321);
-    if (numRings >= 0) {
-      return H3Utils.H3_CORE.kRing(_h3Id, numRings);
-    } else {
-      return Collections.emptyList();
-    }
+    return numRings >= 0 ? getH3Ids(numRings) : Collections.emptyList();
   }
 
   /**
-   * Returns the H3 ids that MIGHT BE partially/fully covered by the circle with the given distance as the radius and a
+   * Returns the H3 ids that MIGHT BE fully/partially covered by the circle with the given distance as the radius and a
    * point within the _h3Id hexagon as the center.
+   * <p>The shortest distance from the center of the center hexagon to the center of a hexagon in the nth ring is
+   * {@code >= 1.5 * n * edgeLength}. Counting the distance from the center to a point in the hexagon, which is up
+   * to the edge length, it is guaranteed that the hexagons in the nth ring are always not fully/partially covered if:
+   * {@code distance < (1.5 * n - 2) * edgeLength}.
    */
-  private List<Long> getPartialMatchH3Ids(double distance) {
+  private List<Long> getPossibleMatchH3Ids(double distance) {
     // NOTE: Add a small delta (0.001) to be conservative
     int numRings = (int) Math.floor((distance / _edgeLength + 2) / 1.5 + 0.001);
+    return getH3Ids(numRings);
+  }
+
+  /**
+   * Returns the H3 ids for the given number of rings around the _h3Id.
+   * IMPORTANT: Throw exception when number of rings is too large because H3 library might send SIGILL for large number
+   * of rings, which can terminate the JVM. Also, the result won't be accurate when the distance is too large. In such
+   * case, the operator will fallback to ExpressionFilterOperator.
+   */
+  private List<Long> getH3Ids(int numRings) {
+    Preconditions.checkState(numRings <= 100, "Expect numRings <= 100, got: %s", numRings);
```

Comment:
Will this fall back to scan-based approach?

Reply:
Yes, as mentioned in the comments